torchtune/models/clip/_tokenizer.py

```diff
@@ -0,0 +1,211 @@
+            assert self.truncate, (
+                "Tokenized text is larger than the maximum sequence length but "
+                "truncate is set to False."
+            )
+            tokens = tokens[: self.max_seq_len]
+            tokens[-1] = self.eot_token
+
+        return tokens
+
+    def decode(self, tokens: List[int]) -> str:
+        """
+        Given a list of token ids, return the decoded text, optionally including special tokens.
+
+        Args:
+            tokens (List[int]): The list of token ids to decode.
+
+        Returns:
+            str: The decoded text.
+        """
+        text = "".join([self.decoder[token] for token in tokens])
+        return (
+            bytearray([self.byte_decoder[c] for c in text])
+            .decode("utf-8", errors="replace")
+            .replace("</w>", " ")
+        )
+
+    def _bpe(self, token):
+        if token in self.cache:
+            return self.cache[token]
+
+        word = tuple(token[:-1]) + (token[-1] + "</w>",)
+        pairs = _get_pairs(word)
+
+        if not pairs:
+            return token + "</w>"
+
+        while True:
+            bigram = min(pairs, key=lambda pair: self.bpe_ranks.get(pair, float("inf")))
+            if bigram not in self.bpe_ranks:
+                break
+            first, second = bigram
+            new_word = []
+            i = 0
+            while i < len(word):
+                try:
+                    j = word.index(first, i)
+                    new_word.extend(word[i:j])
+                    i = j
+                except ValueError:
+                    new_word.extend(word[i:])
+                    break
+
+                if word[i] == first and i < len(word) - 1 and word[i + 1] == second:
+                    new_word.append(first + second)
+                    i += 2
+                else:
+                    new_word.append(word[i])
+                    i += 1
+            word = tuple(new_word)
+            if len(word) == 1:
+                break
+            pairs = _get_pairs(word)
+
+        word = " ".join(word)
+        self.cache[token] = word
+        return word
+
+
+def _bytes_to_unicode():
+    """
+    Returns list of utf-8 byte and a corresponding list of unicode strings.
+    """
+    bs = (
+        list(range(ord("!"), ord("~") + 1))
+        + list(range(ord("¡"), ord("¬") + 1))
+        + list(range(ord("®"), ord("ÿ") + 1))
+    )
+    cs = bs[:]
+    n = 0
+    for b in range(2**8):
+        if b not in bs:
+            bs.append(b)
+            cs.append(2**8 + n)
+            n += 1
+    cs = [chr(n) for n in cs]
+    return dict(zip(bs, cs))
+
+
+def _get_pairs(word):
+    """
+    Return set of symbol pairs in a word.
+    Word is represented as tuple of symbols (symbols being variable-length strings).
+    """
+    pairs = set()
+    prev_char = word[0]
+    for char in word[1:]:
+        pairs.add((prev_char, char))
+        prev_char = char
+    return pairs
+
+
+def _clean_text(text):
```

Comment:
this is needed if we want perfect parity with the HF and OpenAI implementations of the CLIP tokenizer, but I agree that it would be better to remove the `ftfy` dependency and expect the user to clean their dataset. I'll remove this and leave a comment in the code

Reply:
Re-opening this conversation because I ran a parity check on an entire text dataset and I only found one issue: CLIP's text cleaning replaces ’ with ' 

So I re-added a version of the `_clean_text` function that does only this ([link](https://github.com/calvinpelletier/torchtune/blob/5fe86ae50ae0b20867d49a6f168e12a2773afbaf/torchtune/models/clip/_tokenizer.py#L223-L227)). 

This gives us functionally perfect parity with the official implementation without needing to add `ftfy` as a dependency, but it is kinda awkward. Should I remove it and expect the user to clean their dataset themselves? Or should I leave it so that our CLIP tokenizer won't see any strange tokens that it wasn't trained on?

@pbontrager @RdoubleA

Reply:
I'm okay with leaving this in as it's very simple, but I don't think we strictly need it.